Tmedia 836 label counter

JackHowa opened the pull request:
## Ticket

- [TMEDIA-836](https://arcpublishing.atlassian.net/browse/TMEDIA-836)

## Description

use label counter similar to previous implementation

## Acceptance Criteria

- Image counter (i.e. 1 of 15). The counter is labelled on the carousel

## Test Steps

1. Checkout branch - `git checkout TMEDIA-836-label-counter`
2. Run Storybook `npm run storybook`
3. Check out the button storybook documentation. Go to the carousel stories 
4. Notice the label set in the story 
![Screen Shot 2022-04-18 at 10 13 03](https://user-images.githubusercontent.com/5950956/163829582-3d6f0400-a308-4c8e-9709-1e12ef94fadc.png)
5. Click "next" button and notice the label change 
![Screen Shot 2022-04-18 at 10 14 01](https://user-images.githubusercontent.com/5950956/163829659-4faaac44-aee4-4703-a74b-490e40fac607.png)
6. Go full screen by clicking the "Full Screen" button in the Commerce theme 
![Screen Shot 2022-04-18 at 10 15 04](https://user-images.githubusercontent.com/5950956/163829806-633a13c6-bbdc-46a4-a26b-e161d75b75a3.png)
7. Notice how the label text is still visible when this pseudo class of full screen is applied 
![Screen Shot 2022-04-18 at 10 15 54](https://user-images.githubusercontent.com/5950956/163829926-f73b3b1b-78ec-4d14-8711-48e0b55b37f0.png)
8. Exit full screen mode by clicking "Minimize screen" 
9. In the implementation, I've allowed the label text to be modified out of full screen and within. Furthermore, if the user wants to customize the backdrop pseudoelement in full-screen mode, then they could ensure that the text is visible on whatever background color is used with ::backdrop
10. In another story, go to the custom label text. This allows the user to provide a function if available to give internationalized or custom text. This is based off of the engine theme sdk implementation. Notice how this one says "cool images". It will work the same as before
![Screen Shot 2022-04-18 at 10 18 03](https://user-images.githubusercontent.com/5950956/163830222-fb2ece13-6d28-44f3-853d-f7b8932e348a.png)
11. Also, since this is opt-in functionality, no label is shown by default on other stories

## Author Checklist

- [x] Confirmed all the test steps a reviewer will follow above are working.
- [x] Ran this code locally and checked that there are not any unintended side effects. For example, that a CSS selector is scoped only to a particular block.
- [x] Confirmed relevant documentation has been updated/added.
- [x] Add label - **ready for review** when the pull request is ready for someone to begin reviewing

## Reviewer Checklist

_The reviewer of the PR should copy-paste this template into the review comments on review._

- [ ] All GitHub Actions pass
- [ ] Ran the code locally based on the test instructions.
- [ ] Checked Chromatic for Storybook changes, accepted the updates if acceptable
- [ ] Looked to see that the new or changed code has code coverage, specifically. We want the global code coverage to keep on going up with targeted testing.
- [ ] Approve and Add label - **ready to merge** if you are happy with the pull request
- [ ] Want another reviewer? Add the label **additional review**

--------

## Notes 

- Working on ensuring styles can be set for children of a full-screened element. Notice the commerce style allowing the label to be set with one color for not-fullscreen and the other for full-screen. That backdrop can be different colors for different operating systems, browsers
- Added a backdrop for more customization for that as well
- Noticed that slides state variable could track current slide as long as the slides to show does not exceed the slides within. The default value for slides to show is 4 (I might expect 1 for minimum label functionality). 
- Making label opt-in because this is extending the carousel existing functionality

## Questions 

- In terms of internalization and customization, hard-coding "of" is likely not best for the label. Should I pass in the word divider ("of")? or format ("%current off these many %total", "Of %total, you are on %current")? 
-> I can work off the engine-theme-sdk example of setting a default unless a translation is passed in

```
				pageCountPhrase={
							/* istanbul ignore next */ (current, total) =>
								phrases.t("global.gallery-page-count-text", { current, total })
						}

	pageCountPhrase?: (current: number, total: number) => string;

	const ImageCountTextOutput = {
		__html: pageCountPhrase
			? pageCountPhrase(page + 1, galleryElements.length)
			: `<span>Image</span> ${page + 1} of ${galleryElements.length}`,
	};
``` 
- What's the expected behavior for the label with multiple slides showing? With two slides showing, I would expect the global label to say "2 of 5" on first load, right? 
- -> this makes sense to me. This seems to be out of scope for the gallery single slide example

JackHowa committed:
feat: Add top actions grid container

JackHowa committed:
Merge branch 'arc-themes-release-version-2.00' into TMEDIA-836-label-counter

JackHowa committed:
feat: Ensure allowing multiple items within top actions box

JackHowa committed:
feat: Add label for current slide count

JackHowa committed:
docs: Add image counter to ensure that full screen api works as expected for text color

JackHowa committed:
feat: Allow text to be set for within the full screen pseduo class

JackHowa committed:
style: Fix stylelint issue with descending specificity of the image counter label

JackHowa commented on src/components/carousel/_index.scss:
```diff
@@ -69,4 +83,13 @@
 		}
 	}
+
+	&__image-counter-label {
```
updated

JackHowa committed:
feat: Add show label to opt into showing label count

JackHowa committed:
docs: Add storybook for label display

JackHowa committed:
style: Revert accidental update of defaults

JackHowa committed:
style: Revert accidental update to existing storybook

JackHowa committed:
feat: Take in custom label text function

JackHowa committed:
test: Add custom label phrase provider

JackHowa committed:
docs: Update alphabetization for show label

brentmiller1973 commented on src/components/carousel/index.jsx:
```diff
@@ -70,17 +70,20 @@ const Carousel = ({
+	// slidesToShow is a number of slides to show at once
 	const [slide, setSlide] = useState(slidesToShow);
 	const [position, setPosition] = useState(0);
 	const [isFullScreen, setIsFullScreen] = useState(false);
```
I am curious....I tried this out on storybook and it's complete computer fullscreen vs just a "fullscreen" browser window/modal.  Are we sure we want a true full screen presentation?  For example, USNews has a "Full screen" for its gallery, but it's a full browser width and height modal that doesn't take over the user's screen:

![image](https://user-images.githubusercontent.com/833461/163845272-3ea5310a-8c85-42c2-9e4b-eb9d5dbc8d17.png)

Fullscreen computer display seems like it should be reserved for video or games you know, maybe not so much for a gallery.

JackHowa replied:
yep, this is verified from product and Jenae as the desired behavior with full screen api. I put this is in the notes as well as the ticket in a previous pr #64 . @visionbegin  https://arcpublishing.atlassian.net/browse/TMEDIA-834?focusedCommentId=734128

nschubach replied:
We'll probably need a "lightbox mode" or something for that sort of expansion.  But this was for a full screen feature which better fits it's current implementation.

JackHowa replied:
Yeah. I appreciate the question and research from a product side . In my opinion, I prefer using the actual full screen of the window for appreciating the pictures (like videos) as big as possible when I do "expand" a gallery. And Jenae agreed from a product-perspective.

From a technical perspective, I like this because it's using built-in browser technologies. As a result, we save bundle size on package installation (giving us more room for motion libs @visionbegin ). Also, we provide a native experience to users on whatever operating system and browser they are on. Plus we get to use ::backdrop and other pseudoclasses to watch for :fullscreen! Bonus customization 

Also, I wasn't sure where to begin to refactor the existing engine-theme-sdk implementation that has been ported from another library to typescript. In this way, we are starting fresh with our browser support.

brentmiller1973 replied:
Cool, thanks for the explanation....The argument about saving bundle size make a ton of sense!

brentmiller1973 commented on src/components/carousel/index.jsx:
```diff
@@ -195,6 +198,11 @@ const Carousel = ({
 			{...handlers}
 		>
 			<div className={`${COMPONENT_CLASS_NAME}__top-actions`}>
```
Calling this div `*__top-actions` is inappropriate imo.  User may want to move this to the bottom or to the right/left and vertically oriented.  This class needs a more generic name that doesn't suggest placement.

JackHowa replied:
that's a good point. will adjust the naming. maybe controls?

JackHowa replied:
@visionbegin updated -- thanks!

brentmiller1973 replied:
LGTM!

JackHowa committed:
style: Update naming of top-actions to controls per pr feedback

Per https://github.com/WPMedia/arc-themes-components/pull/71#discussion_r852270573

JackHowa committed:
style: Update classname in the component from top-actions to controls per pr feedback on naming specificity

nschubach reviewed (approved):
Looks good with all the adjustments previously noted.  I do wonder if we need to adjust istanbul rules though?  (not required for this, but maybe a future ticket.)

nschubach committed:
Fix classnames
